internal/host/host.go

```diff
@@ -320,23 +339,34 @@ func (m *Manager) RefreshStatus(ctx context.Context, h *models.Host, db *gorm.DB
 	if err != nil {
 		return err
 	}
-	conditions, validationsResults, err := m.rp.preprocess(vc)
+	conditions, newValidationRes, err := m.rp.preprocess(vc)
 	if err != nil {
 		return err
 	}
-	if err = m.reportValidationStatusChanged(ctx, vc, h, validationsResults); err != nil {
+	currentValidationRes, validationsChanged, err := m.checkValidationChanged(ctx, h, newValidationRes)
+	if err != nil {
 		return err
 	}
 	err = m.sm.Run(TransitionTypeRefresh, newStateHost(h), &TransitionArgsRefreshHost{
 		ctx:               ctx,
 		db:                db,
 		eventHandler:      m.eventsHandler,
 		conditions:        conditions,
-		validationResults: validationsResults,
+		validationResults: newValidationRes,
 	})
 	if err != nil {
 		return common.NewApiError(http.StatusConflict, err)
 	}
+	if validationsChanged {
+		// reporting validations change in status needs to compare the new validations results
+		// with the current validations results, therefore, it should be done before updating the DB.
+		m.reportValidationStatusChanged(ctx, vc, h, newValidationRes, currentValidationRes)
+		// the transitions are using the validations info in order to choose the correct
```

Comment:
We never take the validations from db as input to transitions, so it seems the validations may be updated before.  It is even safer because since UpdateHost relies on the state as part of the condition.

Reply:
If we update the DB before the transition, the conditions will be different.